rfc-157-edge-compute.md

```diff
@@ -0,0 +1,236 @@
+
+- Both the Compute@Edge and Lambda@Edge code should live in `govuk-cdn-config`.
+- We should use the [Fastly Terraform provider](https://registry.terraform.io/providers/fastly/fastly/latest/docs) (or possibly the Fastly CLI, if it supports all the features we need) to:
+  - Build the new Compute@Edge service into a WASI bundle and deploy it to Fastly
+  - Configure the two VCL services in the "sandwich"
+    - These will both be architected as entirely new services, i.e. [`www.vcl.erb`](https://github.com/alphagov/govuk-cdn-config/blob/main/vcl_templates/www.vcl.erb) will remain untouched during the migration process (other than perhaps to add A/B testing logic for the new services), and then removed when the migration is complete.
+- A GitHub Actions workflow should be added to `govuk-cdn-config` that, on push to `main`, builds (using e.g. Webpack or esbuild) and packages (e.g. as a series of ZIP archives) the Lambda@Edge code, and commits the built packages to `govuk-aws`.
+  - Committing built Lambda packages to `govuk-aws` is already an [established precedent](https://github.com/alphagov/govuk-aws/blob/main/terraform/lambda/DownloadLogsAnalytics/download_logs_analytics.zip), and allows us to ensure that the Terraform project is always in a deployable state
+- A new Terraform project should be created in `govuk-aws` which:
+  - Creates a new CloudFront distribution, creates the appropriate Lambda functions, and attaches them to the distribution (building upon [this spike](https://github.com/alphagov/govuk-aws/tree/c47b0a265a9f5c6de0b117fe3f48d6c4a56889fb/terraform/lambda/CloudfrontHandlers) in `govuk-aws`, though n.b. I am proposing that the Lambda@Edge code itself should live in `govuk-cdn-config`)
+
+When we are ready to test the new CDN config with live traffic, we could potentially set up A/B testing by adding the new Compute@Edge service and the new CloudFront distribution as (additional) backends of the existing VCL-based service.
+
+## Ownership & roadmap
+
+Platform Reliability would own most of this work.
+
+This project can be divided into a number of stages:
+
+1. Development
+    - We will develop & deploy the new Fastly services and the new CloudFront distribution separately from our existing CDN service.
+    - Both the new Compute@Edge service and the Lambda@Edge handlers should be built together, so that we don't end up painting ourselves into a corner with an architecture that's only suitable for one service.
+    - Both services must have access controls during development (I propose an IP allowlist restricting access to the office & VPN).
+    - The new CloudFront distribution must be added to origin's IP allowlist, so that it is able to make requests to origin. The new Compute@Edge service (or rather, the VCL service behind it in the "sandwich") will [already be covered](https://docs.publishing.service.gov.uk/manual/cdn.html#fastly39s-ip-ranges-and-our-access-controls-on-origin-servers) by our existing allowlist.
```

Comment:
Would prohibit others from making a cloudfront distribution and pointing it at our origin?